File: docs/validation/substrate_math_performance/test-harness/swift/Sources/Harness/Core/SplitMix64.swift
// SplitMix64.swift
//
// SplitMix64 deterministic PRNG for case generation. Same
// algorithm used in `glref-swift-HyperplaneFamily.swift`
// (private there); promoted here so the harness can seed case
// generators reproducibly across languages.
//
// Bit-identical with the Rust harness's `SplitMix64`. Two ports
// seeded with the same u64 produce the same stream.

import Foundation

public struct SplitMix64 {
    private var state: UInt64

    public init(seed: UInt64) {
        self.state = seed
    }

    public mutating func next() -> UInt64 {
        state = state &+ 0x9E3779B97F4A7C15
        var z = state
        z = (z ^ (z >> 30)) &* 0xBF58476D1CE4E5B9
        z = (z ^ (z >> 27)) &* 0x94D049BB133111EB
        return z ^ (z >> 31)
    }

    /// Returns the next `count` u64 values.
    public mutating func nextBytes(count: Int) -> [UInt8] {
        var bytes = [UInt8]()
        bytes.reserveCapacity(count)
        var remaining = count
        while remaining > 0 {
            let w = next()
            for i in 0..<min(8, remaining) {
                bytes.append(UInt8((w >> (i * 8)) & 0xFF))
                remaining -= 1
            }
        }
        return bytes
    }
}

File: docs/validation/substrate_math_performance/test-harness/rust/src/harness/splitmix64.rs
// src/harness/splitmix64.rs
//
// Deterministic PRNG mirroring the Swift harness's SplitMix64.
// Same algorithm, same constants. Two ports seeded with the same
// u64 produce the same stream.

pub struct SplitMix64 {
    state: u64,
}

impl SplitMix64 {
    pub fn new(seed: u64) -> Self {
        Self { state: seed }
    }

    pub fn next(&mut self) -> u64 {
        self.state = self.state.wrapping_add(0x9E37_79B9_7F4A_7C15);
        let mut z = self.state;
        z = (z ^ (z >> 30)).wrapping_mul(0xBF58_476D_1CE4_E5B9);
        z = (z ^ (z >> 27)).wrapping_mul(0x94D0_49BB_1331_11EB);
        z ^ (z >> 31)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deterministic() {
        let mut a = SplitMix64::new(42);
        let mut b = SplitMix64::new(42);
        for _ in 0..10 {
            assert_eq!(a.next(), b.next());
        }
    }

    #[test]
    fn matches_swift_first_three_outputs() {
        // The Swift harness, seed = 42, produces these first
        // three u64 values. CRC32 over their LE concatenation:
        // see the cross-language test in tests/cross_language.rs.
        let mut rng = SplitMix64::new(42);
        let a = rng.next();
        let b = rng.next();
        let c = rng.next();
        // Reference values produced by Swift SplitMix64(seed: 42).
        // SplitMix64 is a public algorithm; these are stable.
        assert_eq!(a, 0xBDD7_3226_2FEB_6E95);
        assert_eq!(b, 0x28EF_E333_B266_F103);
        assert_eq!(c, 0x4752_6757_130F_9F52);
    }
}
